File: test/orderIntentViews.test.js
import assert from "node:assert/strict";
import test from "node:test";
import { getDonorNeighbourhoodWindowMs } from "../src/donorNeighbourhoodWindow.js";
import {
  formatOrderIntentCoordinator,
  formatOrderIntentForRole,
  formatOrderIntentLimited
} from "../src/orderIntentViews.js";

const now = Date.parse("2026-06-02T12:00:00.000Z");

function recordWithPhoto(ageMs, userId = "alice") {
  return {
    id: "oi-1",
    user_id: userId,
    pack_id: "pack-1",
    status: "instructions_copied",
    created_at: new Date(now - ageMs).toISOString(),
    updated_at: new Date(now - ageMs).toISOString(),
    has_reference_photo: true,
    reference_photo_artifact_id: "art-1",
    reference_photo_view_url: "https://cdn/view",
    reference_photo_thumbnail_url: "https://cdn/thumb",
    presets_snapshot: []
  };
}

test("formatOrderIntentLimited strips photo URLs outside neighbourhood window", () => {
  const windowMs = getDonorNeighbourhoodWindowMs();
  const old = recordWithPhoto(windowMs + 1000);
  const formatted = formatOrderIntentLimited(old, now);
  assert.equal(formatted.has_reference_photo, false);
  assert.equal(formatted.reference_photo_view_url, "");
  assert.equal(formatted.reference_photo_thumbnail_url, "");
});

test("formatOrderIntentLimited keeps photo URLs within neighbourhood window", () => {
  const recent = recordWithPhoto(30 * 60 * 1000);
  const formatted = formatOrderIntentLimited(recent, now);
  assert.equal(formatted.reference_photo_view_url, "https://cdn/view");
  assert.equal(formatted.reference_photo_thumbnail_url, "https://cdn/thumb");
});

test("formatOrderIntentForRole uses full view for coordinator", () => {
  const old = recordWithPhoto(getDonorNeighbourhoodWindowMs() + 1000);
  const formatted = formatOrderIntentForRole(old, "coordinator", { nowMs: now });
  assert.equal(formatted.reference_photo_view_url, "https://cdn/view");
});

test("formatOrderIntentCoordinator includes donor_email when known", () => {
  const formatted = formatOrderIntentCoordinator(
    recordWithPhoto(0),
    { alice: "alice@example.com" },
    now
  );
  assert.equal(formatted.user_id, "alice");
  assert.equal(formatted.donor_email, "alice@example.com");
  assert.equal(formatted.initiator_email, "alice@example.com");
});

test("formatOrderIntentLimited omits donor_email", () => {
  const formatted = formatOrderIntentLimited(recordWithPhoto(0), now);
  assert.equal("donor_email" in formatted, false);
});

test("formatOrderIntentLimited keeps coordinates for viewer own intent", () => {
  const formatted = formatOrderIntentLimited(
    {
      ...recordWithPhoto(0),
      location_lat: 12.97,
      location_lng: 80.22
    },
    now,
    "alice"
  );
  assert.equal(formatted.location_lat, 12.97);
  assert.equal(formatted.location_lng, 80.22);
});

test("formatOrderIntentLimited redacts coordinates for other donors", () => {
  const formatted = formatOrderIntentLimited(
    {
      ...recordWithPhoto(0, "bob"),
      location_lat: 12.97,
      location_lng: 80.22
    },
    now,
    "alice"
  );
  assert.equal(formatted.location_lat, null);
  assert.equal(formatted.location_lng, null);
});

File: src/test/java/org/sharingbridge/integration/service/OrderIntentViewsTest.java
package org.sharingbridge.integration.service;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;

import java.time.Instant;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.sharingbridge.integration.geo.DonorNeighbourhoodWindow;

class OrderIntentViewsTest {

    private static final long NOW = Instant.parse("2026-06-02T12:00:00.000Z").toEpochMilli();

    private static Map<String, Object> recordWithPhoto(long ageMs, String userId) {
        Map<String, Object> record = new HashMap<>();
        record.put("id", "oi-1");
        record.put("user_id", userId);
        record.put("pack_id", "pack-1");
        record.put("status", "instructions_copied");
        record.put("created_at", Instant.ofEpochMilli(NOW - ageMs).toString());
        record.put("updated_at", Instant.ofEpochMilli(NOW - ageMs).toString());
        record.put("has_reference_photo", true);
        record.put("reference_photo_artifact_id", "art-1");
        record.put("reference_photo_view_url", "https://cdn/view");
        record.put("reference_photo_thumbnail_url", "https://cdn/thumb");
        record.put("presets_snapshot", List.of());
        return record;
    }

    @Test
    void formatOrderIntentLimitedStripsPhotoUrlsOutsideWindow() {
        long windowMs = DonorNeighbourhoodWindow.getDonorNeighbourhoodWindowMs();
        Map<String, Object> formatted =
                OrderIntentViews.formatOrderIntentLimited(recordWithPhoto(windowMs + 1000, "alice"), NOW, "bob");
        assertEquals(false, formatted.get("has_reference_photo"));
        assertEquals("", formatted.get("reference_photo_view_url"));
        assertEquals("", formatted.get("reference_photo_thumbnail_url"));
    }

    @Test
    void formatOrderIntentLimitedKeepsOwnPhotoUrlsForHistoryList() {
        long windowMs = DonorNeighbourhoodWindow.getDonorNeighbourhoodWindowMs();
        Map<String, Object> formatted =
                OrderIntentViews.formatOrderIntentLimited(recordWithPhoto(windowMs + 1000, "alice"), NOW, "alice");
        assertEquals("https://cdn/view", formatted.get("reference_photo_view_url"));
        assertEquals("https://cdn/thumb", formatted.get("reference_photo_thumbnail_url"));
    }

    @Test
    void formatOrderIntentLimitedKeepsPhotoUrlsWithinWindow() {
        Map<String, Object> formatted =
                OrderIntentViews.formatOrderIntentLimited(
                        recordWithPhoto(30 * 60 * 1000, "alice"), NOW, "");
        assertEquals("https://cdn/view", formatted.get("reference_photo_view_url"));
    }

    @Test
    void formatOrderIntentForRoleUsesFullViewForCoordinator() {
        long windowMs = DonorNeighbourhoodWindow.getDonorNeighbourhoodWindowMs();
        Map<String, Object> formatted =
                OrderIntentViews.formatOrderIntentForRole(
                        recordWithPhoto(windowMs + 1000, "alice"),
                        "coordinator",
                        Map.of(),
                        NOW,
                        "bob");
        assertEquals("https://cdn/view", formatted.get("reference_photo_view_url"));
    }

    @Test
    void formatOrderIntentCoordinatorIncludesDonorEmailWhenKnown() {
        Map<String, Object> formatted =
                OrderIntentViews.formatOrderIntentCoordinator(
                        recordWithPhoto(0, "alice"), Map.of("alice", "alice@example.com"), NOW);
        assertEquals("alice", formatted.get("user_id"));
        assertEquals("alice@example.com", formatted.get("donor_email"));
        assertEquals("alice@example.com", formatted.get("initiator_email"));
    }

    @Test
    void formatOrderIntentLimitedOmitsDonorEmail() {
        Map<String, Object> formatted =
                OrderIntentViews.formatOrderIntentLimited(recordWithPhoto(0, "alice"), NOW, "");
        assertFalse(formatted.containsKey("donor_email"));
    }

    @Test
    void formatOrderIntentLimitedKeepsCoordinatesForViewerOwnIntent() {
        Map<String, Object> record = recordWithPhoto(0, "alice");
        record.put("location_lat", 12.97);
        record.put("location_lng", 80.22);
        Map<String, Object> formatted = OrderIntentViews.formatOrderIntentLimited(record, NOW, "alice");
        assertEquals(12.97, formatted.get("location_lat"));
        assertEquals(80.22, formatted.get("location_lng"));
    }

    @Test
    void formatOrderIntentLimitedRedactsCoordinatesForOtherDonors() {
        Map<String, Object> record = recordWithPhoto(0, "bob");
        record.put("location_lat", 12.97);
        record.put("location_lng", 80.22);
        Map<String, Object> formatted = OrderIntentViews.formatOrderIntentLimited(record, NOW, "alice");
        assertEquals(null, formatted.get("location_lat"));
        assertEquals(null, formatted.get("location_lng"));
    }
}
